core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala

```diff
@@ -0,0 +1,329 @@
+        fail(e.toString)
+      }
+    }
+  }
+
+  /**
+   * Exercises the code in ZkUtils. The goal is mainly
+   * to verify that the behavior of ZkUtils is correct
+   * when isSecure is set to true.
+   */
+  @Test
+  def testZkUtils() {
+    assertTrue(zkUtils.isSecure)
+    for (path <- zkUtils.persistentZkPaths) {
+      zkUtils.makeSurePersistentPathExists(path)
+      if(!path.equals(ZkUtils.ConsumersPath)) {
+        val aclList = (zkUtils.zkConnection.getAcl(path)).getKey
+        assertTrue(aclList.size == 2)
+        for (acl: ACL <- aclList.asScala) {
+          assertTrue(isAclSecure(acl))
+        }
+      }
+    }
+    // Test that can create: createEphemeralPathExpectConflict
+    zkUtils.createEphemeralPathExpectConflict("/a", "")
+    verify("/a")
+    // Test that can create: createPersistentPath
+    zkUtils.createPersistentPath("/b")
+    verify("/b")
+    // Test that can create: createSequentialPersistentPath
+    val seqPath = zkUtils.createSequentialPersistentPath("/c", "")
+    verify(seqPath)
+    // Test that can update: updateEphemeralPath
+    zkUtils.updateEphemeralPath("/a", "updated")
+    val valueA: String = zkUtils.zkClient.readData("/a")
+    assertTrue(valueA.equals("updated"))
+    // Test that can update: updatePersistentPath
+    zkUtils.updatePersistentPath("/b", "updated")
+    val valueB: String = zkUtils.zkClient.readData("/b")
+    assertTrue(valueB.equals("updated"))
+
+    info("Leaving testZkUtils")
+  }
+
+  /**
+   * Tests the migration tool when making an unsecure
+   * cluster secure.
+   */
+  @Test
+  def testZkMigration() {
+    val unsecureZkUtils = ZkUtils(zkConnect, 6000, 6000, false) 
+    try {
+      testMigration(unsecureZkUtils, zkUtils)
+    } finally {
+      unsecureZkUtils.close()
+    }
+  }
+
+  /**
+   * Tests the migration tool when making a secure
+   * cluster unsecure.
+   */
+  @Test
+  def testZkAntiMigration() {
+    val unsecureZkUtils = ZkUtils(zkConnect, 6000, 6000, false)
+    try {
+      testMigration(zkUtils, unsecureZkUtils)
+    } finally {
+      unsecureZkUtils.close()
+    }
+  }
+
+  /**
+   * Tests that the persistent paths cannot be deleted.
+   */
+  @Test
+  def testDelete() {
+    info("zkConnect string: %s".format(zkConnect))
+    for (path <- zkUtils.securePersistentZkPaths) {
+      info("Creating " + path)
+      zkUtils.makeSurePersistentPathExists(path)
+    }    
+    zkUtils.zkConnection.setAcl("/", zkUtils.DefaultAcls, -1)
+    deleteAllUnsecure()
+  }
+
+  /**
+   * Tests that znodes cannot be deleted when the 
+   * persistent paths have children.
+   */
+  @Test
+  def testDeleteRecursive() {
+    info("zkConnect string: %s".format(zkConnect))
+    for (path <- zkUtils.securePersistentZkPaths) {
+      info("Creating " + path)
+      zkUtils.makeSurePersistentPathExists(path)
+      zkUtils.createPersistentPath(path + "/fpjwashere", "")
+    }
+    zkUtils.zkConnection.setAcl("/", zkUtils.DefaultAcls, -1)    
```

Comment:
For `testDeleteRecursive` I want to add children and test that the children also can't be delete. `ZkSecurityMigrator` will only create the parents. But I've done it for `testDelete`, it makes sense for that test case.